### PR TITLE
Merge monaco-editor-wrapper and monaco-editor-react into monaco-languageclient's npm workspace

### LINKED ISSUE #543: Is monaco-languageclient still needed in the future?
This is not a question about if we want to continue to maintain it or not?
The question is whether `monaco-languageclient` is still required as library on its own!? 

In 2023 the actual code of the lib was reduced to a simple `BasicLanguageClient` extension and most of the current lib code is about how to configure `monaco-vscode-api` services. 
With recent discussion in https://github.com/CodinGame/monaco-vscode-api/pull/191 it could even make sense to move away from the current vscode services config approach as it draws in many dependencies (even if conditional) actually not needed by the end-users. This would reduce the code base even further.
This and other things (e.g. how is the lib used) made me wonder if the library is still needed on its own or if it should be integrated elsewhere?

One potential idea is to fuse the code of this lib with [monaco-editor-wrapper](https://github.com/TypeFox/monaco-components/tree/main/packages/monaco-editor-wrapper) as the wrapper already makes usage/integration of `monaco-vscode-api` + `monaco-editor` + `monaco-languageclient` simpler.

Another idea is to move this code to [monaco-vsode-api](https://github.com/CodinGame/monaco-vscode-api) and create a sub-package for the client.

In any case examples should stay here.

What are your thoughts?

### PR DESCRIPTION
**Pull request by kaisalmen**
This PR is now ready for review. A couple of open points should be fixed, but review can start now.

This repository now hosts the following npm packages:
- `monaco-languagclient`; Next release: https://www.npmjs.com/package/monaco-languageclient/v/8.0.0-next.5
- `vscode-ws-jsopnrpc`; Next release: https://www.npmjs.com/package/vscode-ws-jsonrpc/v/3.3.0-next.5
- `monaco-editor-wrapper`; Next release: https://www.npmjs.com/package/monaco-editor-wrapper/v/4.0.0-next.5
- `@typefox/monaco-editor-react`; Next release: https://www.npmjs.com/package/@typefox/monaco-editor-react/v/3.0.0-next.5
- `monaco-languagclient-examples`; Next release: https://www.npmjs.com/package/monaco-languageclient-examples/v/8.0.0-next.5

Important points: 
- The examples of this repository and the ones from https://github.com/TypeFox/monaco-components have been merged.
- Verification examples have been removed from the npm workspace and are now contained in the directory **verify**.
  - The Angular example has been integrated here. The build pipeline verifies they are properly building
  - They have been excluded from main workspace to prevent pollution of the main node_modules (e.g. angular)
- `monaco-editor-workers` is no longer needed and is replaced by `monaco-editor-wrapper/workerFactory` that integrates ideas from monaco-vscode-api demo and is fully customizable/over-writeable.
- Client examples now make use of  `monaco-editor-wrapper` or `@typefox/monaco-editor-react`. This made reduces the amount of boiler plate code.

Left over points required at least required to be implemented before this PR can be merged:

- [x] Browser example should also use the wrapper
- [x] Update all READMEs and linked documentation files
- [x] Check LICENSE files and statements
- [x] Update CHANGELOGs
- [x] Check if existing bugs should be fixed along or if they should be fixed with a bug-fix release
- [ ] The worker based extension host needs to be integrate to allow advanced TypeScript support where needed in the example (wrapperTs and wrapperAdvanced), but this can be deferred.

Will close #543

### CODE REVIEW

**Comment by kaisalmen** [comment]
@CGNonofr I have updated to 2.2.0-next.1 and the font in the command palette is wrong. Should this work without change?
image

`@codingame/monaco-vscode-typescript-basics-default-extension` gives me the following error:
image

and `@codingame/monaco-vscode-typescript-language-features-default-extension` and throws an error:

**Comment by kaisalmen** [comment]
Did I miss changes in the vite configuration? I just checked the your demo config and I didn't see something obvious

**Comment by CGNonofr** [comment]
> Did I miss changes in the vite configuration? I just checked the your demo config and I didn't see something obvious

I'm not aware of anything :thinking: 

The first error looks like the theme service override is not registered

The second one is probably due to the webworker extension host not working (the local process extension host is not able to replace worker urls by real urls)

**Comment by kaisalmen** [comment]
> I'm not aware of anything 🤔

It is consistently using the wrong font. Did you remove to many global css instructions? Interestingly, codicon is fine. In the past, usually problems originated from there.

> The first error looks like the theme service override is not registered

I mixed the classical editor mode of the wrapper with the vscode extensions and therefore the themes override was not loaded. My fault.

> The second one is probably due to the webworker extension host not working (the local process extension host is not able to replace worker urls by real urls)

You have a workaround for that problem in the demo, don't you? Could we generalize that somehow. I need to fix the editor worker issue as well, too.

**Comment by CGNonofr** [comment]
> It is consistently using the wrong font. Did you remove to many global css instructions? Interestingly, codicon is fine. In the past, usually problems originated from there.

There were some removed global style I've added back (with some tweaks to remove properties on the body and so on...)

What do you expect and what do you get?

> You have a workaround for that problem in the demo, don't you? Could we generalize that somehow. I need to fix the editor worker issue as well, too.

Nop, in the demo, the default extensions run in the worker ext host

**Comment by kaisalmen** [comment]
Latest force-push using `@codingame/monaco-vscode-api@2.2.0` fixes the issue [above ](https://github.com/TypeFox/monaco-languageclient/pull/600#issuecomment-1964017110) ⬆️

**Comment by kaisalmen** [comment]
@CGNonofr I am continuing to merge and update the examples (latest code is not yet pushed). Have you seen this error already? Using the standalone editor here with 2.2.1:
image

**Comment by kaisalmen** [comment]
I am not importing any workbench related things and I guess that should not be needed.

**Comment by ls-infra** [comment]
should i put this https://github.com/TypeFox/monaco-languageclient/issues/595 feature on hold until this WIP PR is merged and released ? 
btw, for the language server client runner , is there a way to save the edited value back so that it can submit back to the API?  

thx

**Comment by ls-infra** [comment]
should i put this https://github.com/TypeFox/monaco-languageclient/issues/595 feature on hold until this WIP PR is merged and released ? 
btw, for the language-client-runner , is there a way to save the edited value in code editor back so that it can submit back to the API?   it is related to https://github.com/TypeFox/monaco-languageclient-ng-example/issues/5 ,  i am keen to help to update the angular demo once this two way data binding feature is in place in the wrapper .  

thx

**Comment by kaisalmen** [comment]
> should i put this https://github.com/TypeFox/monaco-languageclient/issues/595 feature on hold until this WIP PR is merged and released ?

Yes, that makes sense. I hope to get this ready finally this week. My own availability will be better again now.

> btw, for the language-client-runner , is there a way to save the edited value in code editor back so that it can submit back to the API? it is related to https://github.com/TypeFox/monaco-languageclient-ng-example/issues/5 , i am keen to help to update the angular demo once this two way data binding feature is in place in the wrapper .

The wrapper allows you already to get access to the editor/model itself. See for example. https://github.com/TypeFox/monaco-languageclient/blob/merge-monaco-edtior-wrapper/packages/wrapper/src/wrapper.ts#L156-L190 This is also put to use in some of the examples.

**Comment by kaisalmen** [comment]
@CGNonofr I have merged your worker loading ideas and monaco-editor-workers (https://github.com/TypeFox/monaco-languageclient/blob/merge-monaco-edtior-wrapper/packages/wrapper/src/workerFactory.ts). It is fully customizable and comes with a preset. Paths and the configuration can be completed overridden.

**Comment by CGNonofr** [comment]
> @CGNonofr I have merged your worker loading ideas and monaco-editor-workers (https://github.com/TypeFox/monaco-languageclient/blob/merge-monaco-edtior-wrapper/packages/wrapper/src/workerFactory.ts). It is fully customizable and comes with a preset. Paths and the configuration can be completed overridden.

Does it work with both webpack and vite? In my experience, Webpack only tries to bundle the worker code it the exact literal syntax is used (`new Worker(new URL(..., import.meta.url), ...)`). Using any intermediate variable breaks it. Or maybe that's a Vite issue?

**Comment by kaisalmen** [comment]
> Does it work with both webpack and vite?

Don't know, yet. Need to still check all verification examples. Will let you know later.

The trick with `monaco-editor-workers` was always to supply a bundle and to work-around Firefox not being able to handle module workers. But, I just realized this need is gone. Firefox support module worker properly. If you supply the bundled worker in a relative location, you are always able to load it.

**Comment by kaisalmen** [comment]
> Webpack only tries to bundle the worker code it the exact literal syntax is used (new Worker(new URL(..., import.meta.url), ...). Using any intermediate variable breaks it. Or maybe that's a Vite issue?

They both don't support. I have verification examples for all cases in place now. I will integrate the Angular example as well. But all verification examples are not part of the npm workspace only the real examples (=prevent dependency pollution of the npm workspace).

**Comment by kaisalmen** [comment]
@CGNonofr Angular verification example is in now as well. The old 1.85.0 broke when creating angular production builds with enabled optimization. This is resolved with the latest version. We are now able to quickly test this here as well. I want to get the final release out of the door this week, but there is some work ahead (docs, some further enhancements to worker handling, updating the browser example...). I will let you know once this is ready for review/testing, probably before doc is ready.

**Comment by kaisalmen** [comment]
@CGNonofr two observations with 3.1.0:
- Locale loading works, but the language in the editor (command palette) is still English. This works in the [deployed demo ](https://codingame.github.io/monaco-vscode-api/?locale=es)
- When I configure a standard monaco theme and do not load the theme service override, the theme works, but I see this error 2-3 seconds after the editor is done loading:
image

I will create a list of required work in the main description above.

**Comment by kaisalmen** [comment]
But once I load the theme service override package isn't monarch support gone or do I understand this incorrectly?

**Comment by CGNonofr** [comment]
> But once I load the theme service override package isn't monarch support gone or do I understand this incorrectly?

There are 2 different things: loading the service override and using the returned services override in the initialize function. You're not supposed to do one without the other

Btw, I think using both TextMate and monach is possible. Tokenization is registered per language and I see no reason why a different tokenizer type per language wouldn't work (outside of the theme token coloration OFC)

**Comment by kaisalmen** [comment]
@CGNonofr Thanks for the explanation. I only import it when I initialize it, at least that is the difference in the wrapper. Maybe I introduced an error/wrong import. Sorry, this PR is too big, but I don't know how to make it smaller. 😊 

Btw, I updated the PR description and release new next releases a couple of minutes ago that reflect the current state/last commit.

**Comment by kaisalmen** [comment]
@CGNonofr browser example is updated and your new code highlighted that my local loading was too late. It is fixed now.

**Comment by kaisalmen** [comment]
@CGNonofr regarding the themes issue. I am positive that textmate and themes services are not loaded. No default vscode default theme is imported, then this error shows up after the 3 seconds timeout. You can check it here: https://github.com/TypeFox/monaco-languageclient/blob/merge-monaco-edtior-wrapper/packages/examples/src/ts/wrapperTs.ts

image

**Comment by CGNonofr** [comment]
> @CGNonofr regarding the themes issue. I am positive that textmate and themes services are not loaded. No default vscode default theme is imported, then this error shows up after the 3 seconds timeout. You can check it here: https://github.com/TypeFox/monaco-languageclient/blob/merge-monaco-edtior-wrapper/packages/examples/src/ts/wrapperTs.ts
> 
> image

You are importing it from editorAppExtended.ts which is imported by the index of the wrapper

**Comment by kaisalmen** [comment]
> You are importing it from editorAppExtended.ts which is imported by the index of the wrapper

Arrghh, you are right. Damn it. Thank you that makes sense

**Comment by kaisalmen** [comment]
⬆️ fixed ✅

**Comment by kaisalmen** [comment]
@CGNonofr I have fixed a couple of bugs. Now it's time for the doc update...

File: packages/examples/src/langium/statemachine/main.ts
Comment (Lemour-sudo) [inline packages/examples/src/langium/statemachine/main.ts:69]: Typo: `loadStatemachin**e**WorkerRegular`

Comment (kaisalmen) [reply packages/examples/src/langium/statemachine/main.ts:69]: Thanks

File: verify/angular/src/app/app.component.ts
Comment (ls-infra) [inline verify/angular/src/app/app.component.ts:32]: is the https://github.com/TypeFox/monaco-languageclient-ng-example repo merged into this monorepo ?
can we add a `console.log` in this component to demo how to read the edited value back from the editor ?

thx,

**Comment by kaisalmen** [comment]
@CGNonofr `@codingame/monaco-vscode-editor-api` is no longer installed as an alias to `monaco-editor`, because I no longer see the necessity. Something like basic-languages from MS monaco-editor cannot be used as they internally reference the editor api and not the package. WDYT about this approach?

**Comment by kaisalmen** [comment]
But wait, this will likely become problematic as user may bring in `monaco-editor` and are not aware of the version clash. 🤔

**Comment by kaisalmen** [comment]
@CGNonofr `monaco-editor` is used as npm dependency designation again.

File: verify/angular/src/app/app.component.ts
Comment (kaisalmen) [reply verify/angular/src/app/app.component.ts:32]: > is the https://github.com/TypeFox/monaco-languageclient-ng-example repo merged into this monorepo ?

Yes, once this PR is merged. The need for the external repo is removed.

> how to read the edited value back from the editor ?

This is done by the react component, see here: https://github.com/TypeFox/monaco-languageclient/blob/merge-monaco-edtior-wrapper/packages/wrapper-react/src/index.tsx#L171-L191 we could also think about moving or having this function directly in the wrapper, but if you need it quickly, you can just use this as a starting point.

**Comment by CGNonofr** [comment]
> Something like basic-languages from MS monaco-editor cannot be used as they internally reference the editor api and not the package. WDYT about this approach?

I don't understand what you mean, can you explain?

**Comment by kaisalmen** [comment]
> I don't understand what you mean, can you explain?

@CGNonofr  This is linked to https://github.com/CodinGame/monaco-vscode-api/issues/363 . monaco-editor is not modular. The basic and enhanced language support can't be used independently of the editor. They have a relative path reference to edtior.api.js and not to the npm package.

This is a strong advantage of the modular npm package approach you chose for `monaco-vscode-api`. If this exists for monaco itself one could combine the two. Currently this is not possible without hacks or copying stuff around.

Btw, I pushed some updates again. Some services util functions need to stay in `monaco-languageclient` otherwise it is no longer really self-standing and I fixed a bug in the enhanced MonacoEnvironment init function I introduced yesterday.

File: packages/wrapper/src/wrapper.ts
Comment (Lemour-sudo) [inline packages/wrapper/src/wrapper.ts:50]: Can this error message be more elaborate that the `codeOriginal` attribute in the config?

Comment (kaisalmen) [reply packages/wrapper/src/wrapper.ts:50]: I will extend the message. Will come with next commit.

**Comment by kaisalmen** [comment]
@CGNonofr README updates and LICENSE unification are done. CHANGELOGs are next.

**Comment by kaisalmen** [comment]
@CGNonofr this is ready for review. Please go ahead.

**Comment by CGNonofr** [comment]
> @CGNonofr this is ready for review. Please go ahead.

There is still a `WIP` commit, am I supposed to review it?

**Comment by kaisalmen** [comment]
> There is still a WIP commit, am I supposed to review it?

A the first one. Nevermind, go ahead

File: packages/client/README.md
Comment (CGNonofr) [inline packages/client/README.md:46]: Once you use?

I'm not sure it's still relevant though since there is no way to even try currently

File: packages/examples/src/utils/localeLoader.ts
Comment (CGNonofr) [inline packages/examples/src/utils/localeLoader.ts:15]: what about the others? (:

**Review by CGNonofr — APPROVED** [review]
didn't spot anything weird but not the easiest PR of my life !

**Comment by kaisalmen** [comment]
@CGNonofr It's rebased and the first to commits on the branch have been merged.

**Review by CGNonofr — APPROVED** [review]

File: packages/examples/src/utils/localeLoader.ts
Comment (kaisalmen) [reply packages/examples/src/utils/localeLoader.ts:15]: It will come...

File: packages/client/README.md
Comment (kaisalmen) [reply packages/client/README.md:46]: It works when manually copying them or using your own monarch grammars. You can run the Langium Grammar DSL example with both configuration and that works fine. 👍

Comment (CGNonofr) [reply packages/client/README.md:46]: Oh right! and those services prevent them from working?

Comment (kaisalmen) [reply packages/client/README.md:46]: That is my understanding, yes.

Comment (kaisalmen) [reply packages/client/README.md:46]: That's what the monaco-vscode-api README says as well, doesn't it? 😆

Comment (CGNonofr) [reply packages/client/README.md:46]: Hum you're right :thinking:

**Pull request merged**
